Introduce a `reserved stock` class and database table to prevent race conditions during checkout

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

We reverted #25708 earlier because checkout would fail if for some reason reserve stock table failed to be created, for example when WordPress DB user do not have create permission on database. 

We now aim to re-introduce this, after building some protection, where this feature will not be enabled if the table fails to be created (see #26454 for database check and #26705 for feature disabling) with the caveat that hold stock functionality will not work. But now checkout will still be successful and the merchant will be able to see a notice which will say that table was not created.

### How to test the changes in this Pull Request:

Same as #25708 .

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enhancement - Updated stock handling to prevent race conditions when orders come in at the same time.

vedanshujain committed:
Revert "Revert "Introduce a `reserved stock` class and database table to prevent race conditions during checkout""

stevepowell2000 commented on includes/class-wc-install.php:
```diff
@@ -946,6 +946,14 @@ private static function get_schema() {
+	`order_id` bigint(20) NOT NULL,
+	`product_id` bigint(20) NOT NULL,
+	`stock_quantity` double NOT NULL DEFAULT 0,
+	`timestamp` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
Suggest a rename here to "created" or "created_at"

mikejolley replied:
timestamp is used in other wc core tables - not we're also using this in blocks so a rename would cause issues for those existing users.

stevepowell2000 commented on includes/class-wc-install.php:
```diff
@@ -946,6 +946,14 @@ private static function get_schema() {
+	`product_id` bigint(20) NOT NULL,
+	`stock_quantity` double NOT NULL DEFAULT 0,
+	`timestamp` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	`expires` datetime NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
suggest default value change to a future timestamp, like 5 minutes

vedanshujain replied:
this is now changed to `0000-00-00 00:00:00` to support MySQL 5.6.

felipe-pita commented:
This will help us alot!

vedanshujain committed:
Alert in status report if base table is not present.

Also add a tool to rerun the check. This needs newer API version to work.

vedanshujain commented:
Note that needs: 
1. https://github.com/woocommerce/woocommerce-rest-api/pull/188
2. https://github.com/woocommerce/woocommerce/pull/26455
3. https://github.com/woocommerce/woocommerce/pull/26454
4. (Incoming)

to be merged before.

mikejolley commented:
@vedanshujain There is a bug due to using CURRENT_TIMESTAMP on a datetime field affecting some installs so we'll probably need to adjust the schema slightly before this is moved to core. I've logged this here https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/2589

vedanshujain committed:
This commit adds two changes for reserved stock functionality:

1. Use '0000-00-00 00:00:00' instead of CURRENT_TIMESTAMP as default value to support MySQL 5.6
2. Return early if DB version is less than 430 because then it would mean that required wc_reserved_stock table might not be present.

rrennick commented:
@vedanshujain When I tried to checkout without the `reserved_stock` table, I got the following error using the cheque payment gateway:

```Not enough units of {{product name}} are available in stock to fulfil this order.```

Once I created the table, checkout proceeded.

peterfabian reviewed (commented):
Looks nice, thanks. Just got 2 questions:

peterfabian commented on includes/class-wc-checkout.php:
```diff
@@ -388,12 +388,30 @@ public function create_order( $data ) {
+			/**
+			 * Action hook fired after an order is created.
+			 *
+			 * @since 4.1.0
```
Should these be updated to 4.3?

vedanshujain replied:
Updated in https://github.com/woocommerce/woocommerce/pull/26395/commits/b8a1a69b2b41c9cde539ff2a223cc35fa3d8714e

vedanshujain commented on src/Checkout/Helpers/ReserveStock.php:
```diff
@@ -0,0 +1,176 @@
+	 */
+	public function release_stock_for_order( \WC_Order $order ) {
+		global $wpdb;
+
```
I think I missed this, good catch! Adding this check. (btw did you meant to post this comment in https://github.com/woocommerce/woocommerce/pull/26705/files)

vedanshujain replied:
Added in https://github.com/woocommerce/woocommerce/pull/26395/commits/d21cc596df72c705d2dccfa41edd4e24245cf6d1

vedanshujain commented on src/Checkout/Helpers/ReserveStock.php:
```diff
@@ -0,0 +1,176 @@
+			"
+			SELECT COALESCE( SUM( stock_table.`stock_quantity` ), 0 ) FROM $wpdb->wc_reserved_stock stock_table
+			LEFT JOIN $wpdb->posts posts ON stock_table.`order_id` = posts.ID
+			WHERE posts.post_status IN ( 'wc-checkout-draft', 'wc-pending' )
```
stocks are for a short while when payment is happening, and in-built order statuses are highly tied during that flow. so if someone wants to customize this, we would also probably need to add filter to change the default status when order is created for example. lets keep it as it is for now, and reconsider if someone reports a good use case. wdyt?

peterfabian replied:
yup, probably too late to add now anyway. Just remembered Gerhard's struggle to add more order statuses, so it's more of a future suggestion than real question...

vedanshujain committed:
Return early in release stock function if stock reservation is disabled.

vedanshujain committed:
Merge pull request #26705 from woocommerce/enhancement/protections-reserved-stock

Port enhancement to reserved stock feature from wc-blocks

mikejolley commented on includes/admin/class-wc-admin-status.php:
```diff
@@ -339,6 +339,29 @@ private static function log_table_bulk_actions() {
+			<?php
+				esc_html_e( 'Missing base tables: ', 'woocommerce' );
+				echo esc_html( implode( ', ', $missing_tables ) );
+				esc_html_e( '. Some WooCommerce functionality may not work as expected.', 'woocommerce' );
```
This might be difficult to translate? Should this use %s instead?

vedanshujain replied:
Good catch. Resolved in https://github.com/woocommerce/woocommerce/pull/26395/commits/34e5d3daa39e9a7475eebeafd3f6617722afeb1f.

vedanshujain committed:
Updated annotations to 4.3

vedanshujain committed:
Use single sentence for easier translations.

peterfabian commented:
Marking as `high impact` to not forget that this should be mentioned in the beta post, because it can create a new table.

peterfabian commented:
Just want to note here that when testing with CompuRacer, I was able to get to -1 stock level. 

Also, after switching back to master, I encountered this weirdness: https://youtu.be/I8QSApDdzww

We might need a follow up on this during the pre-release period.

peterfabian reviewed (approved):
Thanks, this should improve the situation with stock levels quite a bit. Still some smaller problems to iron out, but let's create separate issues for those.

aaqibzahoor commented:
Has this been implemented on normal checkouts of this will only work if we are using Woocommerce blocks?

peterfabian commented:
Hi @aaqibzahoor 

The stock is reserved when the order is created. For Cart & checkout blocks, that is a bit earlier in the process than for shortcodes:

- When you use shortcodes/classic checkout, the order gets created when you click on Place order/Pay via payment gateway.
- When you use blocks for the checkout, an order in status 'draft' gets created on the checkout page, so the stock is reserved one step earlier in the process.

aaqibzahoor commented:
Hi @peterfabian ,

I did a test where the product stock was 1 and  2 users added the product  to the cart at the same time, moved to checkout at the same time and then payed for the product at the same time both the orders went through and the stock was reduced to -1. Which is not very helpful as back orders should not happen. Is it something i am doing wrong or this is the expected behaviour.

Thanks
Aaqib

pduclo commented:
I am a bit confused on when the stock hold is triggered for a customer. 

It sounds like the reserve does not take place until checkout or payment is started.

If there is 1 vacuum available and that vacuum is added to someone's cart, I don't want the next person to be able to add it to their cart. I think the stock should be reserved as soon as it is added to cart. It shouldn't wait for checkout page or payment to start.

morganhvidt commented:
@pduclo Have a look at this diagram for reference https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/main/docs/blocks/stock-reservation.md#how-this-all-fits-into-checkout-block-vs-traditional-checkout

pduclo commented:
Thanks @morganhvidt . That explains it. Not very useful to me until it is like shopping at a physical store. When the item is in the cart it should be an option to reserve that stock. The way it is, it is still a "race" where several customers think they have it in their cart, only to be denied during checkout.

felipe-pita commented:
I`m testing this new feature because I had a lot of problems with negative stock caused by race conditions last year on black friday  :( . I tested with traditional (shortcode) checkout... now it`s better than before, the max negative stock I had on tests was about -7 simulating 100 checkouts processing at same time with same product on cart. I`m testing the Gutenberg checkout to check on and I really liked this feature of creating a draft order as user go to checkout. 

1. Is possible to add this feature on traditional checkout? 
2. I should convert all my payment methods to support this new React checkout, right? Where could I find exemples of how to do that?

thank you!

peterfabian commented:
Hi @felipe-pita ,

> Is possible to add this feature on traditional checkout?
The new database table for stock reservation is part of the 'traditional checkout'. Both Blocks and the Shortcode/Traditional checkout just put the reservation for stock at different points (Blocks when the customer enters the Checkout page, Shortcode checkout after submitting the form). As the link above explains, we use `INSERT INTO ... SELECT ... FOR UPDATE`, which (we hoped) would ensure the query happens as atomic one within a transaction. What database version and engine are you using?

> I should convert all my payment methods to support this new React checkout, right? 
That depends on your use cases. Some products don't work with the Blocks yet, e.g. Subscriptions, or some payment gateways. Please test all your important flows before using the blocks in production, or wait until we include them in WC core experience.

> Where could I find exemples of how to do that?
There is not much you need to actually do. Having the [Blocks plugin](https://wordpress.org/plugins/woo-gutenberg-products-block/) installed is the only prerequisite. Then you can create 2 new pages, add Cart block on one page and Checkout block on the other. As a last step, you set those as Cart and Checkout pages in WooCommerce > Settings > Advanced.

felipe-pita commented:
> The new database table for stock reservation is part of the 'traditional checkout'.

Sorry, but I was referring to "creating a draft order when user go to checkout".

> What database version and engine are you using?

innodb_version 5.7.12

Thank you for your time.

MehbubRashid commented:
i was creating a plugin that shows different stock quantities for different user roles...up until this feature,that means in, wc version less than 4.3 it works perfectly. but after that version, it shows an error.
very sad for me.
![falsdfjlaksdf](https://user-images.githubusercontent.com/34944930/105204354-43fb3980-5b6e-11eb-9c8d-fc3a9af56b60.PNG)
